searches/fibonacci_search.py

```diff
@@ -1,51 +1,110 @@
+
+For manual testing run:
+python3 fibonacci_search.py
 """
+import sys
+from functools import lru_cache
+
+
+@lru_cache()
+def fibonacci(k: int) -> int:
+    """Finds fibonacci number in index k.
 
+    Parameters
+    ----------
+    k :
+        Index of fibonacci.
 
-def fibonacci_search(arr, val):
+    Returns
+    -------
+    int
+        Fibonacci number in position k.
 
+    >>> print(fibonacci(0))
+    0
+    >>> print(fibonacci(2))
+    1
+    >>> print(fibonacci(5))
+    5
+    >>> print(fibonacci(15))
+    610
+    >>> print(fibonacci('a'))
+    Traceback (most recent call last):
+    ValueError: k must be an integer.
+    >>> print(fibonacci(-5))
+    Traceback (most recent call last):
+    ValueError: k integer must be greater or equal to zero.
     """
-    >>> fibonacci_search([1,6,7,0,0,0], 6)
+    if not isinstance(k, int):
+        raise ValueError("k must be an integer.")
+    if k < 0:
+        raise ValueError("k integer must be greater or equal to zero.")
+    if k == 0:
+        return 0
+    elif k == 1:
+        return 1
+    else:
+        return fibonacci(k - 1) + fibonacci(k - 2)
+
+
+def fibonacci_search(arr: list, val: int) -> int:
+    """A pure Python implementation of a fibonacci search algorithm.
+
+    Parameters
+    ----------
+    arr
+        List of sorted elements.
+    val
+        Element to search in list.
+
+    Returns
+    -------
+    int
+        The index of the element in the array.
+        -1 if the element is not found.
+
+    >>> print(fibonacci_search([4, 5, 6, 7], 4))
+    0
+    >>> print(fibonacci_search([4, 5, 6, 7], -10))
+    -1
+    >>> print(fibonacci_search([-18, 2], -18))
+    0
+    >>> print(fibonacci_search([5], 5))
+    0
+    >>> print(fibonacci_search(['a', 'c', 'd'], 'c'))
     1
-    >>> fibonacci_search([1,-1, 5, 2, 9], 10)
+    >>> print(fibonacci_search(['a', 'c', 'd'], 'f'))
     -1
+    >>> print(fibonacci_search([], 1))
+    -1
+    >>> print(fibonacci_search([.1, .4 , 7], .4))
+    1
     >>> fibonacci_search([], 9)
-    0
+    -1
     """
-    fib_N_2 = 0
-    fib_N_1 = 1
-    fibNext = fib_N_1 + fib_N_2
-    length = len(arr)
-    if length == 0:
-        return 0
-    while fibNext < len(arr):
-        fib_N_2 = fib_N_1
-        fib_N_1 = fibNext
-        fibNext = fib_N_1 + fib_N_2
-    index = -1
-    while fibNext > 1:
-        i = min(index + fib_N_2, (length - 1))
-        if arr[i] < val:
-            fibNext = fib_N_1
-            fib_N_1 = fib_N_2
-            fib_N_2 = fibNext - fib_N_1
-            index = i
-        elif arr[i] > val:
-            fibNext = fib_N_2
-            fib_N_1 = fib_N_1 - fib_N_2
-            fib_N_2 = fibNext - fib_N_1
-        else:
-            return i
-    if (fib_N_1 and index < length - 1) and (arr[index + 1] == val):
-        return index + 1
-    return -1
+    n = len(arr)
+    # Find m such that F_m >= n where F_i is the i_th fibonacci number.
+    m = next(x for x in range(sys.maxsize ** 10) if fibonacci(x) >= n)
```

Comment:
`sys.maxsize` is already a huge number. Can you tell me why is this necessary?
And please use descriptive variable names instead of the single letter names as mentioned in CONTRIBUTING.md. Apply the changes wherever the variables were used.

Reply:
`sys.maxsize` is a simple way to represent that the generator produced can go up to infinity. Probably is not the best way to find it.

It is equal to do something like this.

```
i = 0
while True:
    if fibbonacci(i) >= n:
        m = i
        break
    i += 1
```

Both are loops that can go up to infinity. But I do not know which one of them is clearer.

Reply:
I commented it inside the conflict, but I will paste the message here, because I think there it is not easy to find.

`sys.maxsize` is a simple way to represent that the generator produced can go up to infinity. Probably is not the best way to find it.

It is equal to do something like this.

```
i = 0
while True:
    if fibbonacci(i) >= n:
        m = i
        break
    i += 1
```

Both are loops that can go up to infinity. But I do not know which one of them is clearer.

Reply:
As far as I understand, you are finding the index of a fibonacci number which is greater than the length of the input array, right? If that's the case then why are you using a generator just to find one value?

It seems like I probably found a bug:
```python
print(fibonacci_search(list(range(100)), 63))

...
Traceback (most recent call last):
  File "fibonacci_search.py", line 117, in <module>
    print(fibonacci_search(list(range(100)), 63))
  File "fibonacci_search.py", line 106, in fibonacci_search
    offset += fibonacci(fibb_k - 2)
  File "fibonacci_search.py", line 49, in fibonacci
    raise ValueError("k integer must be greater or equal to zero.")
ValueError: k integer must be greater or equal to zero.
```

Reply:
> As far as I understand, you are finding the index of a fibonacci number which is greater than the length of the input array, right? If that's the case then why are you using a generator just to find one value?

Because when I use, next of a generator it will only find the first value of the iterator. And the first value of that generator is the value which I am looking for.

I think is OK but the code in this repo should be clean and understandable. Do you think that it is better for me to change it to something easier? like the snippet above.

Reply:
I think the above mentioned snippet is more readable (with descriptive variable names). Also, take a look at the error produced in the above comment.

Reply:
> I think the above mentioned snippet is more readable (with descriptive variable names).
Agreed, I change it for the sake of clarity.

>  Also, take a look at the error produced in the above comment.
Thank you for pointing that out, it is a special case of the fibonacci search. I fixed it and added your example and many more with big list as doctests.